# On Data Cleaning and Exploration of Ebay Car Sales

In this project, the main objective is to do data cleaning and explore a used cars database available on [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). The data contains 50,000 records of used-car sales on Ebay. We will explore several columns that are possibly useful for future analysis and clean those columns. 


# 1. Importing Data

We first start by importing the data using pandas and exploring its columns.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv',encoding='Latin-1')
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We see that the dataset has 20 columns, some of which contain null values. In this project, we limit our interest to the following columns, and a description of a column is copied from its [documentation](https://www.kaggle.com/orgesleka/used-cars-database/data) for reference.
<br><br>
`seller`: private or dealer
<br>`offerType`: type of listing
<br>`nrOfPictures`: number of pictures in the ad
<br>`odometer`: how many kilometers the car has driven
<br>`price`: the price on the ad to sell the car
<br>`dateCrawled`: when this ad was first crawled, all field-values are taken from this date
<br>`dateCreated`: the date for which the ad at ebay was created
<br>`lastSeen`: when the crawler saw this ad last online
<br>`yearOfRegistration`: at which year the car was first registered
<br>`brand`: the brand of the car

## 2. General Cleaning

We first rename the columns to snakecase, e.g, from `abtest` to `ab_test`, for consistency and easy reference. Also, columns with either long names or ambiguous names are replaced by more intuitive ones.

In [3]:
autos.columns = ['date_crawled','name','seller','offer_type','price','ab_test','vehicle_type','registration_year','gearbox','power_ps','model','odometer','registration_month','fuel_type','brand','unrepaired_damage','ad_created','no_of_pictures','postal_code','last_seen']

autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Looking at the description of the data, we find that some columns have a small number of unique values. For example, the `seller` and `offer_type` columns have only 2 unique values. We explore these two columns further. Moreover, the `no_of_pictures` column also seems to contain no useful information at all.

In [4]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [5]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [6]:
autos["no_of_pictures"].value_counts()

0    50000
Name: no_of_pictures, dtype: int64

We see that the `seller` and `offer_type` columns have mostly the same values for every row, and the `no_of_pictures` row has all zero values. Thus, these columns tell us little information and may not be useful for analysis. We may consider dropping these columns. 

In [7]:
autos = autos.drop(["no_of_pictures", "seller", "offer_type"], axis=1)

Next, we notice that `price` and `odometer` columns contain useful numerical data, so we change the format of these columns to numeric types for further analysis. In particular, we want to change from `'$5,000'` to `5000`, and from `'150,000km'` to `150000`. To denote the change, we also rename the `odometer` column to `odometer_km`.

In [8]:
autos['price'] = (autos["price"].str.replace('$','')
                                .str.replace(',','')
                                .astype(float)
                 )

autos['odometer'] = (autos['odometer'].str.replace('km','')
                                        .str.replace(',','')
                                        .astype(float))
autos.rename(columns = {'odometer':'odometer_km'}, inplace= True)

# 3. Cleaning price and odometer

We take a closer look at the `price` and `odometer` columns.

In [9]:
autos["odometer_km"].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

We see that the majority of car listed have high milleges, i.e., above 150,000km. 

In [10]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Looking at the basic statistics of the price, we see that some cars are listed with price as low as 0, and as high as 100 million dollars. They seem to be the outliers of the dataset. We investigate them more closely.

In [11]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
Name: price, dtype: int64

In [12]:
autos["price"].value_counts().sort_index(ascending=True).head(15)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
Name: price, dtype: int64

We see that we have 1,421 car listings at \\$0, accounted for roughly 2.8\% of the whole dataset. Since cars listed at \\$0 does not bring useful information, we will simply delete these rows. On the other hand, we have few cars listed at very high price. They seem to be outlier and we will delete the listings whose price is more than \\$350,000, as these listings are not really realistic. However, listings at \\$1 is not problematic since Ebay is an auction site, so users can start the base price at as low as \\$1.

In [13]:
autos = autos[autos["price"].between(1,350001)]

# 4. Cleaning date and time columns

Next, we will look at columns containing date and time information. Let us look at unique values for the `data_crawled` column together with its frequency normalized to between 0 and 1, sorted by the  dates in ascending order.

In [15]:
# selecting the first 10 characters representing the date
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

We notice that the data contains the records between March 5th and April 7th 2016, and rougly the same amount of data is recorded every day. 

We move on to take a look at the `ad_created` column.

In [16]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

We see that most of the listings are close to the data collecting period, i.e., March 5th to April 7th, 2016. However, we do have 6 listings from 2015.

Now let's investigate the `last_seen` column. `last_seen` records the date when the listing is last seen on Ebay. The posts are no longer there most probably because the cars have been sold. 

In [17]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

We notice that the number of `last_seen` suddenly goes up in the last three days. This is likely the results of closing data collection period.

What's about the `registration_year` column?

In [19]:
autos["registration_year"].value_counts(normalize=True, dropna=False).sort_index()

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000082
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000082
1957    0.000041
1958    0.000082
1959    0.000124
1960    0.000474
1961    0.000124
1962    0.000082
1963    0.000165
1964    0.000247
          ...   
2000    0.064985
2001    0.054278
2002    0.051189
2003    0.055575
2004    0.055657
2005    0.060455
2006    0.054978
2007    0.046886
2008    0.045609
2009    0.042932
2010    0.032719
2011    0.033419
2012    0.026974
2013    0.016535
2014    0.013652
2015    0.008072
2016    0.025121
2017    0.028663
2018    0.009678
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000082
5911    0.000021
6200    0.000021
8888    0.0000

We see that there is something clearly wrong about some of the values, as we cannot have the registration year too low as 1000 or too high as 8888. Also, since the data is collected in 2016, a car cannot have the registration year in 2017 and beyond. Because the percentage of these outlier listings is not high, we will simply delete those records.

In [20]:
autos = autos[autos["registration_year"].between(1900,2016)]

# 5. Cleaning brand

As for the next step, we'd like to look at the `brand` column.

In [23]:
autos["brand"].value_counts(normalize=True, dropna=False).sort_values()

lada              0.000578
lancia            0.001071
rover             0.001328
trabant           0.001392
daewoo            0.001500
jaguar            0.001564
saab              0.001649
land_rover        0.002099
subaru            0.002142
jeep              0.002271
daihatsu          0.002506
dacia             0.002635
chrysler          0.003513
chevrolet         0.005698
suzuki            0.005934
porsche           0.006127
alfa_romeo        0.006641
kia               0.007069
honda             0.007840
mitsubishi        0.008226
mini              0.008762
volvo             0.009147
sonstige_autos    0.009811
hyundai           0.010025
toyota            0.012703
citroen           0.014010
smart             0.014160
mazda             0.015188
nissan            0.015274
skoda             0.016409
seat              0.018273
fiat              0.025642
peugeot           0.029841
renault           0.047150
ford              0.069900
audi              0.086566
mercedes_benz     0.096463
o

We notice that the top 5 brands that account for more than 60% of the listings are German brands, with Volkswagen being the most popular one; its numer of listing nearly doubles the following number of BMV listings. 

# 6. Concluding Remarks

In this project, we cleaned and explored several columns of interest in the Ebay car sales dataset. Based on this cleaned dataset, one can continue and perform various analysis on the price, the brand, and the date of the listings. 